README.md
# ForkWithFlags

[![Mix Tests](https://github.com/tylerbarker/fork_with_flags/actions/workflows/test.yml/badge.svg?branch=master)](https://github.com/tylerbarker/fork_with_flags/actions/workflows/test.yml?query=branch%3Amaster)
[![Code Quality](https://github.com/tylerbarker/fork_with_flags/actions/workflows/quality.yml/badge.svg?branch=master)](https://github.com/tylerbarker/fork_with_flags/actions/workflows/quality.yml?query=branch%3Amaster)  
[![Hex.pm](https://img.shields.io/hexpm/v/fork_with_flags.svg)](https://hex.pm/packages/fork_with_flags)
[![hexdocs.pm](https://img.shields.io/badge/docs-1.11.1-brightgreen.svg)](https://hexdocs.pm/fork_with_flags/1.11.1/ForkWithFlags.html)
[![Hex.pm Downloads](https://img.shields.io/hexpm/dt/fork_with_flags)](https://hex.pm/packages/fork_with_flags)
[![License](https://img.shields.io/hexpm/l/fork_with_flags.svg)](https://github.com/tylerbarker/fork_with_flags/blob/master/LICENSE.txt)
[![ElixirWeekly](https://img.shields.io/badge/featured-ElixirWeekly-8e5ab5.svg)](https://elixirweekly.net/issues/43)
[![ElixirCasts](https://img.shields.io/badge/featured-ElixirCasts-ff931e.svg)](https://elixircasts.io/feature-flags)

## A note on the fork

This is a temporary fork of [tompave](https://github.com/tompave)'s excellent [fun_with_flags](https://github.com/tompave/fun_with_flags) library. It seems that the original library is not actively maintained at the moment, so I've created this fork. For my case, specifically to integrate unmerged changes which add support for SQLite as a database adapter. That being said I'm also happy to review and accept other contributions to the library while the original repository remains in stasis.

I in no way take credit for, or attribute myself to, this library. All that goes to [tompave](https://github.com/tompave), whom I'm happy to hand this work back over to if and whenever the time comes.

Thanks ✌️

## Switching from fun_with_flags

I tried my best to keep the original naming of `fun_with_flags`, `FunWithFlags`, etc, but I was unable to do so in a way that made the compiler happy, and I've done this to publish on Hex so `fork_with_flags` it is!

A few things you'll have to do to integrate the fork...

### 0. Don't change the table names

To ease migration to the fork, **I've intentionally not changed anything about the default migrations and table names and Redis DB names**. The default table name remains `fun_with_flags_toggles` (`fun_with_flags` for Redis). So, when it comes to replacing `fun_with_flags` all over the place, keep the `ecto_table_name: "fun_with_flags_toggles"` config intact - unless you're already using a custom name.

### 1. Switch packages in mix.exs

Unlock the dependencies:
```shell
mix deps.unlock fun_with_flags
mix deps.unlock fun_with_flags_ui # if necessary
```

Then update the package names and install, re-compile, etc:

```elixir
{:fork_with_flags, "~> 1.11.2"},
{:fork_with_flags_ui, "~> 0.8.2"}, # if necessary
```

### 2. Replace fun_with_flags -> fork_with_flags

Remember: Leave the `:ecto_table_name` configuration option intact unless you're using a custom table name already.

You'll find `fun_with_flags` all around in places like `included_applications: [...]`, `dialyzer: [...]`, etc. Replace those with `fun_with_flags`.

Also ensure your replace all covers `Endpoint.subscribe("fork_with_flags_changes")` to become `Endpoint.subscribe("fork_with_flags_changes")`.

### 3. Replace FunWithFlags -> ForkWithFlags

Usage of the library itself will also need to be replaced, from `FunWithFlags` to `ForkWithFlags`.

That's about it!

## README

ForkWithFlags, a fork on the Elixir feature flag library [fun_with_flags](https://github.com/tompave/fun_with_flags).

If you're reading this on the [GitHub repo](https://github.com/tylerbarker/fork_with_flags), keep in mind that this readme refers to the `master` branch. For the latest version released on Hex, please check [the readme published with the docs](https://hexdocs.pm/fork_with_flags/readme.html).

---

ForkWithFlags is an OTP application that provides a 2-level storage to save and retrieve feature flags, an Elixir API to toggle and query them, and a [web dashboard](#web-dashboard) as control panel.

It stores flag information in Redis or a relational DB (PostgreSQL, MySQL, or SQLite - with Ecto) for persistence and synchronization across different nodes, but it also maintains a local cache in an ETS table for fast lookups. When flags are added or toggled on a node, the other nodes are notified via PubSub and reload their local ETS caches.

## Content

* [What's a Feature Flag](#whats-a-feature-flag)
* [Usage](#usage)
  - [Gate Priority and Interactions](#gate-priority-and-interactions)
  - [Boolean Gate](#boolean-gate)
  - [Actor Gate](#actor-gate)
  - [Group Gate](#group-gate)
  - [Percentage of Time Gate](#percentage-of-time-gate)
  - [Percentage of Actors Gate](#percentage-of-actors-gate)
  - [Clearing a Feature Flag's Rules](#clearing-a-feature-flags-rules)
* [Web Dashboard](#web-dashboard)
* [Origin](#origin)
* [So, caching, huh?](#so-caching-huh)
* [To Do](#to-do)
* [Installation](#installation)
* [Configuration](#configuration)
  - [Persistence Adapters](#persistence-adapters)
    - [Ecto Multi-tenancy](#ecto-multi-tenancy)
  - [PubSub Adapters](#pubsub-adapters)
* [Extensibility](#extensibility)
  - [Custom Persistence Adapters](#custom-persistence-adapters)
* [Application Start Behaviour](#application-start-behaviour)
* [Testing](#testing)
* [Development](#development)

## What's a Feature Flag?

Feature flags, or feature toggles, are boolean values associated to a name. They should be used to control whether some application feature is enabled or disabled, and they are meant to be modified at runtime while an application is running. This is usually done by the people who control the application.

In their simplest form, flags can be toggled on and off globally. More advanced rules or "gates" allow a fine grained control over their status. For example, it's possible to toggle a flag on and off for specific entities or for groups.

The goal is to have more granular and precise control over what is made available to which users, and when.
A common use case, in web applications, is to enable a functionality without the need to deploy or restart the server, or to enable it only for internal users to test it before rolling it out to everyone. Another scenario is the ability to quickly disable a functionality if it's causing problems.
They can also be used to implement a simple authorization system, for example to an admin area.


## Usage

ForkWithFlags has a simple API to query and toggle feature flags. Most of the time, you'll call `ForkWithFlags.enabled?/2` with the name of the flag and optional arguments.

Different kinds of toggle gates are supported:

* **Boolean**: globally on and off.
* **Actors**: on or off for specific structs or data. The `ForkWithFlags.Actor` protocol can be implemented for types and structs that should have specific rules. For example, in web applications it's common to use a `%User{}` struct or equivalent as an actor, or perhaps the current country of the request.
* **Groups**: on or off for structs or data that belong to a category or satisfy a condition. The `ForkWithFlags.Group` protocol can be implemented for types and structs that belong to groups for which a feature flag can be enabled or disabled. For example, one could implement the protocol for a `%User{}` struct to identify administrators.
* **%-of-Time**: globally on for a percentage of the time. It ignores actors and groups. Mutually exclusive with the %-of-actors gate.
* **%-of-Actors**: globally on for a percentage of the actors. It only applies when the flag is checked with a specific actor and is ignored when the flag is checked without actor arguments. Mutually exclusive with the %-of-time gate.

Boolean, Actor and Group gates can express either an enabled or disabled state. The percentage gates can only express an enabled state, as disabling something for a percentage of time or actors is logically equivalent to enabling it for the complementary percentage.

### Gate Priority and Interactions

The priority order is from most to least specific: `Actors > Groups > Boolean > Percentage`, and it applies to both enabled and disabled gates.

For example, a disabled group gate takes precedence over an enabled boolean (global) gate for the entities in the group, and a further enabled actor gate overrides the disabled group gate for a specific entity. When an entity belongs to multiple groups with conflicting toggle status, the disabled group gates have precedence over the enabled ones. The percentage gates are checked last, if present, and they're only checked if no other gate is enabled.

As another example, a flag can have a disabled boolean gate and a 50% enabled %-of-actors gate. When the flag is checked with an actor, it has a (deterministic, consistent and repeatable) 50% chance to be enabled, but when checked without an actor argument it will always be disabled. If we add to the flag a disabled actor gate and an enabled group gate, the flag will be always disabled for the actor, always enabled for any other actor matching the group, have a 50% change to be enabled for any other actor, and always be disabled when checked without actor arguments. If, then, we replace the 50%-of-actors gate with a 50%-of-time gate, the flag will be always disabled for the actor, always enabled for any other actor matching the group, and have a 50% chance to be enabled for any other actor or when checked without an actor argument.

### Boolean Gate

The boolean gate is the simplest one. It's either enabled or disabled, globally. It's also the gate with the second lowest priority (it can mask the percentage gates). If a flag is undefined, it defaults to be globally disabled.

```elixir
ForkWithFlags.enabled?(:cool_new_feature)
false

{:ok, true} = ForkWithFlags.enable(:cool_new_feature)

ForkWithFlags.enabled?(:cool_new_feature)
true

{:ok, false} = ForkWithFlags.disable(:cool_new_feature)

ForkWithFlags.enabled?(:cool_new_feature)
false
```

### Actor Gate

This allows you to enable or disable a flag for one or more entities. For example, in web applications it's common to use a `%User{}` struct or equivalent as an actor, or perhaps the data used to represent the current country for an HTTP request. This can be useful to showcase a work-in-progress feature to someone, to gradually rollout a functionality by country, or to dynamically disable some features in some contexts.

Actor gates take precedence over the others, both when they're enabled and when they're disabled. They can be considered as toggle overrides.

In order to be used as an actor, an entity must implement the `ForkWithFlags.Actor` protocol. This can be implemented for custom structs or literally any other type.


```elixir
defmodule MyApp.User do
  defstruct [:id, :name]
end

defimpl ForkWithFlags.Actor, for: MyApp.User do
  def id(%{id: id}) do
    "user:#{id}"
  end
end

bruce = %MyApp.User{id: 1, name: "Bruce"}
alfred = %MyApp.User{id: 2, name: "Alfred"}

ForkWithFlags.Actor.id(bruce)
"user:1"
ForkWithFlags.Actor.id(alfred)
"user:2"

defimpl ForkWithFlags.Actor, for: Map do
  def id(%{actor_id: actor_id}) do
    "map:#{actor_id}"
  end

  def id(map) do
    map
    |> inspect()
    |> (&:crypto.hash(:md5, &1)).()
    |> Base.encode16
    |> (&"map:#{&1}").()
  end
end

ForkWithFlags.Actor.id(%{actor_id: "bar"})
"map:bar"
ForkWithFlags.Actor.id(%{foo: "bar"})
"map:E0BB5BA6873E3AC34B0B6928190C1F2B"
```

With the protocol implemented, actors can be used with the library functions:

```elixir
{:ok, true} = ForkWithFlags.enable(:restful_nights)
{:ok, false} = ForkWithFlags.disable(:restful_nights, for_actor: bruce)
{:ok, true} = ForkWithFlags.enable(:batmobile, for_actor: bruce)

ForkWithFlags.enabled?(:restful_nights)
true
ForkWithFlags.enabled?(:batmobile)
false

ForkWithFlags.enabled?(:restful_nights, for: alfred)
true
ForkWithFlags.enabled?(:batmobile, for: alfred)
false

ForkWithFlags.enabled?(:restful_nights, for: bruce)
false
ForkWithFlags.enabled?(:batmobile, for: bruce)
true
```

Actor identifiers must be globally unique binaries. Since supporting multiple kinds of actors is a common requirement, all the examples use the common technique of namespacing the IDs:

```elixir
defimpl ForkWithFlags.Actor, for: MyApp.User do
  def id(user) do
    "user:#{user.id}"
  end
end

defimpl ForkWithFlags.Actor, for: MyApp.Country do
  def id(country) do
    "country:#{country.iso3166}"
  end
end
```

### Group Gate

Group gates are similar to actor gates, but they apply to a category of entities rather than specific ones. They can be toggled on or off for the _name of the group_ instead of a specific term.

Group gates take precedence over boolean gates but are overridden by actor gates.

Group names can be binaries or atoms. Atoms are supported for retro-compatibility with versions `<= 0.9` and binaries are therefore preferred. In fact, atoms are internally converted to binaries and are then stored and later retrieved as binaries.

The semantics to determine which entities belong to which groups are application specific.
Entities could have an explicit list of groups they belong to, or the groups could be abstract and inferred from some other attribute. For example, an `:employee` group could comprise all `%User{}` structs with an email address matching the company domain, or an `:admin` group could be made of all users with `%User{admin: true}`.

In order to be affected by a group gate, an entity should implement the `ForkWithFlags.Group` protocol. The protocol automatically falls back to a default `Any` implementation, which states that any entity belongs to no group at all. This makes it possible to safely use "normal" actors when querying group gates, and to implement the protocol only for structs and types for which it matters.

The protocol can be implemented for custom structs or literally any other type.


```elixir
defmodule MyApp.User do
  defstruct [:email, admin: false, groups: []]
end

defimpl ForkWithFlags.Group, for: MyApp.User do
  def in?(%{email: email}, "employee"), do: Regex.match?(~r/@mycompany.com$/, email)
  def in?(%{admin: is_admin}, "admin"), do: !!is_admin
  def in?(%{groups: list}, group_name), do: group_name in list
end

elisabeth = %MyApp.User{email: "elisabeth@mycompany.com", admin: true, groups: ["engineering", "product"]}
ForkWithFlags.Group.in?(elisabeth, "employee")
true
ForkWithFlags.Group.in?(elisabeth, "admin")
true
ForkWithFlags.Group.in?(elisabeth, "engineering")
true
ForkWithFlags.Group.in?(elisabeth, "marketing")
false

defimpl ForkWithFlags.Group, for: Map do
  def in?(%{group: group_name}, group_name), do: true
  def in?(_, _), do: false
end

ForkWithFlags.Group.in?(%{group: "dumb_tests"}, "dumb_tests")
true
```

With the protocol implemented, actors can be used with the library functions:

```elixir
ForkWithFlags.disable(:database_access)
ForkWithFlags.enable(:database_access, for_group: "engineering")

ForkWithFlags.enabled?(:database_access)
false
ForkWithFlags.enabled?(:database_access, for: elisabeth)
true
```

### Percentage of Time Gate

%-of-time gates are similar to boolean gates, but they allow to enable a flag for a percentage of the time. In practical terms, this means that a percentage of the `enabled?()` calls for a flag will return true, regardless of the presence of an actor argument.

When a %-of-time gate is checked a [pseudo-random number is generated](http://erlang.org/doc/man/rand.html#uniform-1) and compared with the percentage value of the gate. If the result of the random roll is lower than the gate's percentage value, the gate is considered enabled. So, at the risk of stating the obvious and for the sake of clarity, a 90% gate is enabled more often than a 10% gate.

%-of-time gates are useful to gradually introduce alternative code paths that either have the same effects of the old ones, or don't have effects visible to the users. This last point is important, because with a %-of-time gate the application will behave differently on a pseudo-random basis.

A good use case for %-of-time gates is to safely test the correctness or performance and load characteristics of an alternative implementation of a functionality.

For example:

```elixir
ForkWithFlags.clear(:alternative_implementation)
ForkWithFlags.enable(:alternative_implementation, for_percentage_of: {:time, 0.05})

def foo(bar) do
  if ForkWithFlags.enabled?(:alternative_implementation) do
    new_foo(bar)
  else
    old_foo(bar)
  end
end
```

The %-of-time gate is incompatible and mutually exclusive with the %-of-actors gate, and it replaces it when it gets set. While there are ways to make them work together, it would needlessly overcomplicate the priority rules.

### Percentage of Actors Gate

%-of-actors gates are similar to the %-of-time gates, but instead of using a pseudo-random chance they calculate the actor scores using a deterministic, consistent and repeatable function that factors in the flag name. At a high level:

```elixir
actor
|> ForkWithFlags.Actor.id()
|> sha256_hash(flag_name)
|> hash_to_percentage()
```

Since the scores depend on both the actor ID and the flag name, they're guaranteed to always be the same for each actor-flag combination. At the same time, the same actor will have different scores for different flags, and each flag will have a uniform distribution of scores for all the actors.

Just like for the %-of-time gates, an actor's score is compared with the gate's percentage value and, if lower, the gate will result enabled.

A practical example, based on the `ForkWithFlags.Actor` protocol set up from the previous sections:

```elixir
defmodule MyApp.User do
  defstruct [:id, :name]
end

defimpl ForkWithFlags.Actor, for: MyApp.User do
  def id(%{id: id}) do
    "user:#{id}"
  end
end

frodo  = %MyApp.User{id: 1, name: "Frodo Baggins"}
sam    = %MyApp.User{id: 2, name: "Samwise Gamgee"}
pippin = %MyApp.User{id: 3, name: "Peregrin Took"}
merry  = %MyApp.User{id: 4, name: "Meriadoc Brandybuck"}

ForkWithFlags.Actor.Percentage.score(frodo, :pipeweed)
0.8658294677734375
ForkWithFlags.Actor.Percentage.score(sam, :pipeweed)
0.68426513671875
ForkWithFlags.Actor.Percentage.score(pippin, :pipeweed)
0.510528564453125
ForkWithFlags.Actor.Percentage.score(merry, :pipeweed)
0.2617645263671875

{:ok, true} = ForkWithFlags.enable(:pipeweed, for_percentage_of: {:actors, 0.60})

ForkWithFlags.enabled?(:pipeweed, for: frodo)
false
ForkWithFlags.enabled?(:pipeweed, for: sam)
false
ForkWithFlags.enabled?(:pipeweed, for: pippin)
true
ForkWithFlags.enabled?(:pipeweed, for: merry)
true

{:ok, true} = ForkWithFlags.enable(:pipeweed, for_percentage_of: {:actors, 0.685})

ForkWithFlags.enabled?(:pipeweed, for: sam)
true

ForkWithFlags.Actor.Percentage.score(pippin, :pipeweed)
0.510528564453125
ForkWithFlags.Actor.Percentage.score(pippin, :mushrooms)
0.6050872802734375
ForkWithFlags.Actor.Percentage.score(pippin, :palantir)
0.144073486328125
```


Once a %-of-actors gate has been defined for a flag, the same actor will always see the same result (unless its actor or group gates are set, or the flag gets globally enabled). Also, this means that as long the percentage value of the gate will increase and never decrease, actors for which the gate has been enabled will always see it enabled.

This is ideal to gradually roll out new functionality to users.

For example, in a Phoenix application:

```elixir
ForkWithFlags.clear(:new_design)
ForkWithFlags.enable(:new_design, for_percentage_of: {:actors, 0.2})
ForkWithFlags.enable(:new_design, for_group: "beta_testers")


defmodule MyPhoenixApp.MyView do
  use MyPhoenixApp, :view

  def render("my_template.html", assigns) do
    if ForkWithFlags.enabled?(:new_design, for: assigns.user) do
      render("new_template.html", assigns)
    else
      render("old_template.html", assigns)
    end
  end
end
```

The %-of-actors gate is incompatible and mutually exclusive with the %-of-time gate, and it replaces it when it gets set. While there are ways to make them work together, it would needlessly overcomplicate the priority rules.

### Clearing a Feature Flag's Rules

Sometimes enabling or disabling a gate is not what you want, and removing that gate's rules would be more correct. For example, if you don't need anymore to explicitly enable or disable a flag for an actor or for a group, and the default state should be used instead, clearing the gate is the right choice.

More examples:


CHANGELOG.md
# Changelog

## v1.11.2

* Mistake in README.

## v1.11.1

* README updates.
* Update broadcast channel to `"fork_with_flags_changes"`.

## v1.11.0

* Add support for SQLite with the [`ecto_sqlite3`](https://hex.pm/packages/ecto_sqlite3) adapter.

## v1.10.1

* Redis notifications adapter: minor internal changes to what data is passed to the supervisor in the child spec. ([pull/148](https://github.com/tompave/fun_with_flags/pull/148))
* Redis notifications adapter: fix an issue that would arise when configuring both a Redis URL string and key-word options (this config API was introduced in v1.10.0): the config would work for the Redis persistence adapter, but not for the Redis notifications adapter. (Thanks [iamvery](https://github.com/iamvery), [pull/149](https://github.com/tompave/fun_with_flags/pull/149))

## v1.10.0

* Add support for Elixir 1.14. Drop support for Elixir 1.11. Elixir >= 1.12 is now required. Dropping support for older versions of Elixir simply means that this package is no longer tested with them in CI, and that compatibility issues are not considered bugs.
* Drop support for Erlang/OTP 22, and Erlang/OTP >= 23 is now required. Dropping support for older versions of Erlang/OTP simply means that this package is not tested with them in CI, and that no compatibility issues are considered bugs.
* Ecto persistence adapter: ForkWithFlags will now pass a custom option when using the [Ecto Repo Query API](https://hexdocs.pm/ecto/3.8.4/Ecto.Repo.html#query-api): `[fun_with_flags: true]`. This is done to make it easier to identify ForkWithFlags queries when working with Ecto customization hooks, e.g. the [`Ecto.Repo.prepare_query/3` callback](https://hexdocs.pm/ecto/3.8.4/Ecto.Repo.html#c:prepare_query/3). (Thanks [SteffenDE](https://github.com/SteffenDE), [pull/143](https://github.com/tompave/fun_with_flags/pull/143))
* Redis persistence adapter: added support to configure Redis with a `{"redis URL", [...kw opts]}` tuple, [as is supported in Redix itself](https://hexdocs.pm/redix/1.2.0/Redix.html#start_link/2). (Thanks [iamvery](https://github.com/iamvery), [pull/145](https://github.com/tompave/fun_with_flags/pull/145))

## v1.9.0

* Drop support for Elixir 1.10. Elixir >= 1.11 is now required. Dropping support for older versions of Elixir simply means that this package is no longer tested with them in CI, and that compatibility issues are not considered bugs.
* Relax supported versions of `postgrex` to allow `~> 0.16`.
* Use [`Application.compile_env/3`](https://hexdocs.pm/elixir/1.13.3/Application.html#compile_env/3) to read the persistence config at compile time, which is used to configure the DB table name when using the Ecto persistence adapter (among other things). This fixes another instance of the issue where users of the package would change the config after compilation and observe unexpected inconsistencies and errors. ([pull/130](https://github.com/tompave/fun_with_flags/pull/130))
* Redis adapters: add support to configure [Redis Sentinel](https://redis.io/docs/manual/sentinel/). Please see the [Redix docs](https://github.com/whatyouhide/redix/tree/v1.1.5#redis-sentinel) for more details. (Thanks [parkdoyeon](https://github.com/parkdoyeon), [pull/107](https://github.com/tompave/fun_with_flags/pull/107).)
* More precise conditional checks when deciding whether Ecto files should be compiled. ([pull/140](https://github.com/tompave/fun_with_flags/pull/140))
* Improved documentation for running the package in a custom supervision tree, when using releases. (Thanks [zaid](https://github.com/zaid), [pull/139](https://github.com/tompave/fun_with_flags/pull/139))

## v1.8.1

* Lock `postgrex` dependency to `< 0.16`. Version `0.16` requires Elixir 1.11 ([changelog](https://github.com/elixir-ecto/postgrex/blob/master/CHANGELOG.md#v0160-2022-01-23)) and it doesn't compile with Elixit 1.10, which ForkWithFlags still supports.

## v1.8.0

* Add support for Elixir 1.13. Drop support for Elixir 1.9. Elixir >= 1.10 is now required. Dropping support for older versions of Elixir simply means that this package is no longer tested with them in CI, and that compatibility issues are not considered bugs.
* Removed all uses of [`defdelegate/2`](https://hexdocs.pm/elixir/1.13.0/Kernel.html#defdelegate/2). They caused some references to configured modules (that can change according to the config) to be reified at compile time, which lead to unexpected behaviour. They've been replaced with plain old function definitions that do the same job. (Thanks [connorlay](https://github.com/connorlay), [pull/111](https://github.com/tompave/fun_with_flags/pull/111).)
* Local dev: Update the config for the library to use [`Config`](https://hexdocs.pm/elixir/1.13.0/Config.html) instead of the deprecated [`Mix.Config`](https://hexdocs.pm/mix/1.13.0/Mix.Config.html). For the avoidance of doubt: this has no effect when using the package in your projects, because the `config/*.exs` files are not present in the bundles downloaded from Hex.pm.
* Use [`Application.compile_env/3`](https://hexdocs.pm/elixir/1.13.0/Application.html#compile_env/3) to read the cache configuration at compile-time, which is used to define a module attribute (therefore, set at compile-time). That part of the config is compiled into a module attribute for performance reasons, and it has been a long standing issue because users of the package would get confused by their config changes not being reflected in an already compiled application ([link to relevant section in previous version of the readme](https://github.com/tompave/fun_with_flags/tree/v1.7.0#configuration-changes-have-no-effect-in-mix_envdev)). Now, if the relevant configuration changes, users will get a [clear error](https://github.com/elixir-lang/elixir/blob/v1.10/CHANGELOG.md#tracking-of-compile-time-configuration).

mix.exs
defmodule ForkWithFlags.Mixfile do
  use Mix.Project

  @source_url "https://github.com/tylerbarker/fork_with_flags"
  @version "1.11.2"

  def project do
    [
      app: :fork_with_flags,
      version: @version,
      elixir: "~> 1.12",
      elixirc_paths: elixirc_paths(Mix.env()),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      package: package(),
      docs: docs(),
      aliases: aliases(),
      dialyzer: dialyzer()
    ]
  end

  def application do
    # Specify extra applications you'll use from Erlang/Elixir
    [extra_applications: extra_applications(Mix.env()), mod: {ForkWithFlags.Application, []}]
  end

  defp extra_applications(:test), do: local_extra_applications()
  defp extra_applications(:dev), do: local_extra_applications()
  defp extra_applications(_), do: [:logger]

  # When working locally with the Ecto adapter, start the ecto_sql
  # and postgrex applications. They're not started automatically
  # because they're optional, I think.
  #
  # Also start the Phoenix PubSub application if that notification
  # adapter is configured.
  #
  defp local_extra_applications do
    apps =
      if System.get_env("PERSISTENCE") == "ecto" do
        [:logger, :ecto, :ecto_sql, :postgrex]
      else
        [:logger, :redix]
      end

    apps =
      if System.get_env("PUBSUB_BROKER") == "phoenix_pubsub" do
        [:phoenix_pubsub | apps]
      else
        apps
      end

    apps
  end

  defp deps do
    [
      {:redix, "~> 1.0", optional: true},
      {:ecto_sql, "~> 3.0", optional: true},
      {:ecto_sqlite3, "~> 0.9", optional: true, only: [:dev, :test]},
      {:postgrex, "~> 0.16", optional: true, only: [:dev, :test]},
      {:myxql, "~> 0.2", optional: true, only: [:dev, :test]},
      {:phoenix_pubsub, "~> 2.0", optional: true},
      {:mock, "~> 0.3", only: :test},
      {:ex_doc, "~> 0.21", only: :dev, runtime: false},
      {:credo, "~> 1.6", only: :dev, runtime: false},
      {:dialyxir, "~> 1.0", only: :dev, runtime: false}
    ]
  end

  defp dialyzer do
    [
      # Add optional dependencies to avoid "unknown_function" warnings.
      plt_add_apps: [:redix, :ecto, :ecto_sql, :phoenix_pubsub]
    ]
  end

  defp aliases do
    [
      {:"test.all", [&run_all_tests/1]},
      {:"test.phx", [&run_tests__redis_pers__phoenix_pubsub/1]},
      {:"test.ecto.postgres", [&run_tests__ecto_pers_postgres__phoenix_pubsub/1]},
      {:"test.ecto.mysql", [&run_tests__ecto_pers_mysql__phoenix_pubsub/1]},
      {:"test.ecto.sqlite", [&run_tests__ecto_pers_sqlite__phoenix_pubsub/1]},
      {:"test.redis", [&run_tests__redis_pers__redis_pubsub/1]}
    ]
  end

  # Runs all the test configurations.
  # If any fails, exit with status code 1, so that this can properly fail in CI.
  #
  defp run_all_tests(arg) do
    tests = [
      &run_tests__redis_pers__redis_pubsub/1,
      &run_integration_tests__redis_pers__redis_pubsub__no_cache/1,
      &run_tests__redis_pers__phoenix_pubsub/1,
      &run_integration_tests__redis_pers__phoenix_pubsub__no_cache/1,
      &run_tests__ecto_pers_postgres__phoenix_pubsub/1,
      &run_integration_tests__ecto_pers_postgres__phoenix_pubsub__no_cache/1,
      &run_tests__ecto_pers_mysql__phoenix_pubsub/1,
      &run_integration_tests__ecto_pers_mysql__phoenix_pubsub__no_cache/1,
      &run_tests__ecto_pers_sqlite__phoenix_pubsub/1,
      &run_integration_tests__ecto_pers_sqlite__phoenix_pubsub__no_cache/1
    ]

    exit_codes =
      case System.get_env("CI") do
        "true" ->
          tests
          |> Enum.map(fn test_fn -> _run_test_with_retries(3, 500, fn -> test_fn.(arg) end) end)

        _ ->
          tests |> Enum.map(fn test_fn -> test_fn.(arg) end)
      end

    if Enum.any?(exit_codes, &(&1 != 0)) do
      require Logger
      Logger.error("Some test configuration did not pass.")
      exit({:shutdown, 1})
    end
  end

  # Because some tests are flaky in CI.
  #
  defp _run_test_with_retries(attempts, sleep_ms, test_fn) when attempts > 0 do
    IO.puts(
      "---\nRunning a test task with retries. Attempts left: #{attempts}, sleep ms: #{sleep_ms}.\n---"
    )

    case test_fn.() do
      # Successful run, simply return the status.
      0 ->
        0

      _ ->
        :timer.sleep(sleep_ms)
        remaining = attempts - 1
        IO.puts("Test failed. Retries left: #{remaining}.")
        _run_test_with_retries(remaining, sleep_ms, test_fn)
    end
  end

  defp _run_test_with_retries(_, _, _) do
    IO.puts("---\nAll retries failed. Returning exit code 1.\n---")
    1
  end

  # Run the tests with Redis as persistent store and Redis PubSub as broker.
  #
  # Cache enabled, force re-compilation.
  #
  defp run_tests__redis_pers__redis_pubsub(arg) do
    Mix.shell().cmd(
      "mix test --color --force --exclude phoenix_pubsub --exclude ecto_persistence #{arg}",
      env: [
        {"CACHE_ENABLED", "true"}
      ]
    )
  end

  # Runs the integration tests only.
  # Cache disabled, Redis as persistent store and Redis PubSub as broker.
  #
  defp run_integration_tests__redis_pers__redis_pubsub__no_cache(arg) do
    Mix.shell().cmd(
      "mix test --color --force --only integration #{arg}",
      env: [
        {"CACHE_ENABLED", "false"}
      ]
    )
  end

  # Run the tests with Redis as persistent store and Phoenix.PubSub as broker.
  #
  defp run_tests__redis_pers__phoenix_pubsub(arg) do
    Mix.shell().cmd(
      "mix test --color --force --exclude redis_pubsub --exclude ecto_persistence #{arg}",
      env: [
        {"CACHE_ENABLED", "true"},
        {"PUBSUB_BROKER", "phoenix_pubsub"}
      ]
    )
  end

  # Runs the integration tests only.
  # Cache disabled, Redis as persistent store and Phoenix.PubSubas broker.
  #
  defp run_integration_tests__redis_pers__phoenix_pubsub__no_cache(arg) do
    Mix.shell().cmd(
      "mix test --color --force --only integration #{arg}",
      env: [
        {"CACHE_ENABLED", "false"},
        {"PUBSUB_BROKER", "phoenix_pubsub"}
      ]
    )
  end

  # Run the tests with Ecto+PostgreSQL as persistent store and Phoenix.PubSub as broker.
  #
  defp run_tests__ecto_pers_postgres__phoenix_pubsub(arg) do
    Mix.shell().cmd(
      "mix test --color --force --exclude redis_pubsub --exclude redis_persistence #{arg}",
      env: [
        {"CACHE_ENABLED", "true"},
        {"PUBSUB_BROKER", "phoenix_pubsub"},
        {"PERSISTENCE", "ecto"},
        {"RDBMS", "postgres"}
      ]
    )
  end

  # Run the tests with Ecto+MySQL as persistent store and Phoenix.PubSub as broker.
  #
  defp run_tests__ecto_pers_mysql__phoenix_pubsub(arg) do
    Mix.shell().cmd(
      "mix test --color --force --exclude redis_pubsub --exclude redis_persistence #{arg}",
      env: [
        {"CACHE_ENABLED", "true"},
        {"PUBSUB_BROKER", "phoenix_pubsub"},
        {"PERSISTENCE", "ecto"},
        {"RDBMS", "mysql"}
      ]
    )
  end

  # Run the tests with Ecto+SQLite as persistent store and Phoenix.PubSub as broker.
  #
  defp run_tests__ecto_pers_sqlite__phoenix_pubsub(arg) do
    Mix.shell().cmd(
      "mix test --color --force --exclude redis_pubsub --exclude redis_persistence #{arg}",
      env: [
        {"CACHE_ENABLED", "true"},
        {"PUBSUB_BROKER", "phoenix_pubsub"},
        {"PERSISTENCE", "ecto"},
        {"RDBMS", "sqlite"}
      ]
    )
  end

  # Runs the integration tests only.
  # Cache disabled, Ecto+PostgreSQL as persistent store and Phoenix.PubSub as broker.
  #
  defp run_integration_tests__ecto_pers_postgres__phoenix_pubsub__no_cache(arg) do
    Mix.shell().cmd(
      "mix test --color --force --only integration #{arg}",
      env: [
        {"CACHE_ENABLED", "false"},
        {"PUBSUB_BROKER", "phoenix_pubsub"},
        {"PERSISTENCE", "ecto"},
        {"RDBMS", "postgres"}
      ]
    )
  end

  # Runs the integration tests only.
  # Cache disabled, Ecto+MySQL as persistent store and Phoenix.PubSub as broker.
  #
  defp run_integration_tests__ecto_pers_mysql__phoenix_pubsub__no_cache(arg) do
    Mix.shell().cmd(
      "mix test --color --force --only integration #{arg}",
      env: [
        {"CACHE_ENABLED", "false"},
        {"PUBSUB_BROKER", "phoenix_pubsub"},
        {"PERSISTENCE", "ecto"},
        {"RDBMS", "mysql"}
      ]
    )
  end

  # Runs the integration tests only.
  # Cache disabled, Ecto+SQLite as persistent store and Phoenix.PubSub as broker.
  #
  defp run_integration_tests__ecto_pers_sqlite__phoenix_pubsub__no_cache(arg) do
    Mix.shell().cmd(
      "mix test --color --force --only integration #{arg}",
      env: [
        {"CACHE_ENABLED", "false"},
        {"PUBSUB_BROKER", "phoenix_pubsub"},
        {"PERSISTENCE", "ecto"},
        {"RDBMS", "sqlite"}
      ]
    )
  end

  defp elixirc_paths(:test), do: ["lib", "test/support", "dev_support"]
  defp elixirc_paths(:dev), do: ["lib", "dev_support"]
  defp elixirc_paths(_), do: ["lib"]

  defp description do
    """
    ForkWithFlags, a fork of the flexible and fast feature toggle library for Elixir.
    """
  end

  defp package do
    [
      maintainers: [
        "Tommaso Pavese",
        "Tyler Barker"
      ],
      licenses: [
        "MIT"
      ],
      links: %{
        "GitHub" => @source_url,
        "Changelog" => "#{@source_url}/blob/master/CHANGELOG.md"
      }
    ]
  end

  defp docs do
    [
      extras: ["README.md", "CHANGELOG.md"],
      main: "ForkWithFlags",
      source_url: @source_url,
      source_ref: "v#{@version}"
    ]
  end
end
